core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala

```diff
@@ -54,22 +53,108 @@ class ShardedHttpExecutionServiceCluster (readMembers: Map[ExecutionServiceId,Ex
+  //
+  // if it's a subworkflow (or sub-sub-workflow, etc) it's not present in the Rawls DB and we don't know which
+  // execution service shard has processed it.  executionServiceKeyOpt = None.  Query all execution service shards
+  // for the workflow to learn its submission association and which shard processed it.  In practice, one shard does
+  // everything except for some older workflows on shard 2.  Revisit this if that changes!
+
+  val SUBMISSION_ID_KEY = "workbench-submission-id"
+
+  private def parseSubWorkflowIdsFromMetadata(metadata: String): Seq[String] = {
+
+  //    Workflow metadata has this structure:
+  //
+  //    {
+  //      "calls" : {
+  //        "foo.bar" : [
+  //          {
+  //            "subWorkflowId" : "69581e76-2eb0-4179-99b9-958d210ebc4b", ...
+  //          }, ...
+  //        ], ...
+  //      },  ...
+  //    }
+
+
+    import spray.json._
+    import spray.json.DefaultJsonProtocol._
+
+    for {
+      callsObj <- JsonParser(metadata).asJsObject().getFields("calls")
+      call <- callsObj.asJsObject().fields.values
+      shard <- call.asInstanceOf[JsArray].elements
+      id <- shard.asJsObject().getFields("subWorkflowId")
+    } yield id.convertTo[String]
+  }
+
+  // parse subworkflow IDs from the parent workflow's metadata and label each of them with the Submission ID
+  // so they can be found by future workflow metadata queries
+
+  private def labelSubWorkflowsWithSubmissionId(submissionId: String, executionServiceId: ExecutionServiceId, parentWorkflowMetadata: JsObject, userInfo: UserInfo): Future[Seq[ExecutionServiceLabelResponse]] = {
+    // traverse = stop on "first" failure and propagate
+    Future.traverse(parseSubWorkflowIdsFromMetadata(parentWorkflowMetadata.compactPrint)) { subWorkflowId =>
+      getMember(executionServiceId).patchLabels(subWorkflowId, userInfo, Map(SUBMISSION_ID_KEY -> submissionId))
+    }
+  }
+
+  def callLevelMetadata(submissionId: String, workflowId: String, executionServiceKeyOpt: Option[ExecutionServiceId], userInfo: UserInfo): Future[JsObject] = {
+    val executionServiceKeyFut = executionServiceKeyOpt match {
+      case Some(executionServiceId) =>
+        // single-workflow or top-level work case: workflow found in DB and confirmed to be a member of this submission
+        // we have the execution service key from the DB, so query the correct execution service
+        Future.successful(executionServiceId)
+
+      case _ =>
+        // we don't have the execution service key because the workflow is not in the DB.  It might be a subworkflow.
+        // query all execution services for Workflow labels and search for a Submission match
+
+        val executionServiceLabelMap = readMembers.map { case (executionServiceId, executionServiceDao) =>
+          executionServiceDao.getLabels(workflowId, userInfo) map { labelResponse =>
+            (executionServiceId, labelResponse.labels)
+          }
+        }
```

Comment:
given what we know about cromwell sharding's legacy and especially the timing of when subworkflows were well-supported, we *expect* that all subworkflows will be on the cromwell submitServer, and none will be on the read-only extra shard. Therefore, this is making a probably-extraneous API call to the read-only shard for every subworkflow.

What if we changed this to query the expected (submit) shard first, and only if nothing is returned from that shard do we query the extra (read) shard? In the expected case this will cut down on API calls; in the corner case it will cause a delay due to the serial order of the queries.

Reply:
OK, makes sense